File: packages/open-design-cli/src/utils/variant-loader.ts
import { readFile, fileExists } from './file-utils.js';
import { join } from 'path';
import YAML from 'yaml';

export interface SkillConfig {
  name: string;
  engine?: string;
  tools: string[];
  config: Record<string, any>;
}

export interface VariantConfig {
  name: string;
  base_role: string;
  additional_skills: string[];
  skill_overrides: Record<string, Record<string, any>>;
  skill_priority: Record<string, number>;
  description: string;
  tags: string[];
}

export interface RoleConfig {
  name: string;
  skills: Record<string, SkillConfig>;
  description: string;
  variants: string[];
}

export class VariantLoader {
  private rolesDir: string;
  private skillsDir: string;
  private roleCache: Map<string, RoleConfig>;
  private variantCache: Map<string, VariantConfig>;

  constructor(rolesDir: string, skillsDir: string) {
    this.rolesDir = rolesDir;
    this.skillsDir = skillsDir;
    this.roleCache = new Map();
    this.variantCache = new Map();
  }

  async loadRole(roleName: string): Promise<RoleConfig> {
    if (this.roleCache.has(roleName)) {
      return this.roleCache.get(roleName)!;
    }

    const rolePath = join(this.rolesDir, roleName, 'role.md');
    if (!(await fileExists(rolePath))) {
      throw new Error(`Role not found: ${roleName}`);
    }

    // Parse role.md to extract skills
    // This is a simplified implementation
    // In production, you'd parse the markdown properly
    const roleConfig: RoleConfig = {
      name: roleName,
      skills: {},
      description: '',
      variants: [],
    };

    // Load variants if they exist
    const variantsPath = join(this.rolesDir, roleName, 'variants.yaml');
    if (await fileExists(variantsPath)) {
      const variants = await this.loadVariantsYaml(variantsPath);
      roleConfig.variants = variants.map(v => v.name);
      for (const variant of variants) {
        this.variantCache.set(variant.name, variant);
      }
    }

    this.roleCache.set(roleName, roleConfig);
    return roleConfig;
  }

  private async loadVariantsYaml(variantsPath: string): Promise<VariantConfig[]> {
    const content = await readFile(variantsPath);
    const data = YAML.parse(content);

    const variants: VariantConfig[] = [];
    for (const variantData of data.variants || []) {
      const variant: VariantConfig = {
        name: variantData.name,
        base_role: variantData.base_role,
        additional_skills: variantData.additional_skills || [],
        skill_overrides: variantData.skill_overrides || {},
        skill_priority: variantData.skill_priority || {},
        description: variantData.description || '',
        tags: variantData.tags || [],
      };
      variants.push(variant);
    }

    return variants;
  }

  async loadVariant(variantName: string): Promise<RoleConfig> {
    // Find variant in cache
    let variant = this.variantCache.get(variantName);
    if (!variant) {
      // Try to find it by scanning all roles
      variant = await this.findVariant(variantName);
      if (!variant) {
        throw new Error(`Variant not found: ${variantName}`);
      }
    }

    // Load base role
    const baseRole = await this.loadRole(variant.base_role);

    // Merge skills
    const mergedSkills = this.mergeSkills(
      baseRole.skills,
      variant.additional_skills
    );

    // Apply skill overrides
    for (const [skillName, overrideConfig] of Object.entries(variant.skill_overrides)) {
      if (skillName in mergedSkills) {
        const skill = mergedSkills[skillName];
        for (const [key, value] of Object.entries(overrideConfig)) {
          (skill as any)[key] = value;
        }
      }
    }

    // Resolve conflicts based on priority
    const resolvedSkills = this.resolveConflicts(
      mergedSkills,
      variant.skill_priority
    );

    // Create new role config
    const variantRole: RoleConfig = {
      name: variant.name,
      skills: resolvedSkills,
      description: variant.description,
      variants: [],
    };

    return variantRole;
  }

  private async findVariant(variantName: string): Promise<VariantConfig | undefined> {
    const fs = await import('fs/promises');
    const { readdir } = fs;
    
    try {
      const roleDirs = await readdir(this.rolesDir, { withFileTypes: true });
      for (const roleDir of roleDirs) {
        if (roleDir.isDirectory()) {
          const variantsPath = join(this.rolesDir, roleDir.name, 'variants.yaml');
          if (await fileExists(variantsPath)) {
            const variants = await this.loadVariantsYaml(variantsPath);
            for (const variant of variants) {
              if (variant.name === variantName) {
                this.variantCache.set(variant.name, variant);
                return variant;
              }
            }
          }
        }
      }
    } catch (error) {
      // Directory doesn't exist or can't be read
    }
    
    return undefined;
  }

  private mergeSkills(
    baseSkills: Record<string, SkillConfig>,
    additionalSkillNames: string[]
  ): Record<string, SkillConfig> {
    const merged: Record<string, SkillConfig> = { ...baseSkills };

    for (const skillName of additionalSkillNames) {
      // Simplified - just create a basic skill config
      const skillConfig: SkillConfig = {
        name: skillName,
        tools: [],
        config: {},
      };
      merged[skillName] = skillConfig;
    }

    return merged;
  }

  private resolveConflicts(
    skills: Record<string, SkillConfig>,
    priorities: Record<string, number>
  ): Record<string, SkillConfig> {
    // If no priorities, return as-is
    if (Object.keys(priorities).length === 0) {
      return skills;
    }

    // Sort skills by priority
    const sortedSkills = Object.entries(skills).sort(
      ([nameA], [nameB]) => {
        const priorityA = priorities[nameA] || 0;
        const priorityB = priorities[nameB] || 0;
        return priorityB - priorityA; // Higher priority first
      }
    );

    const resolved: Record<string, SkillConfig> = {};
    for (const [name, skill] of sortedSkills) {
      resolved[name] = skill;
    }

    return resolved;
  }

  async listVariants(roleName?: string): Promise<string[]> {
    if (roleName) {
      const role = await this.loadRole(roleName);
      return role.variants;
    }

    // List all variants
    const allVariants: string[] = [];
    const fs = await import('fs/promises');
    const { readdir } = fs;
    
    try {
      const roleDirs = await readdir(this.rolesDir, { withFileTypes: true });
      for (const roleDir of roleDirs) {
        if (roleDir.isDirectory()) {
          const variantsPath = join(this.rolesDir, roleDir.name, 'variants.yaml');
          if (await fileExists(variantsPath)) {
            const variants = await this.loadVariantsYaml(variantsPath);
            allVariants.push(...variants.map(v => v.name));
          }
        }
      }
    } catch (error) {
      // Directory doesn't exist or can't be read
    }

    return allVariants;
  }

  async validateVariant(variantName: string): Promise<string[]> {
    const errors: string[] = [];

    try {
      let variant = this.variantCache.get(variantName);
      if (!variant) {
        variant = await this.findVariant(variantName);
        if (!variant) {
          errors.push(`Variant not found: ${variantName}`);
          return errors;
        }
      }

      // Validate base role exists
      const baseRolePath = join(this.rolesDir, variant.base_role, 'role.md');
      if (!(await fileExists(baseRolePath))) {
        errors.push(`Base role not found: ${variant.base_role}`);
      }

      // Validate additional skills exist
      for (const skillName of variant.additional_skills) {
        const skillPath = join(this.skillsDir, skillName, 'SKILL.md');
        if (!(await fileExists(skillPath))) {
          errors.push(`Skill not found: ${skillName}`);
        }
      }

      // Validate skill overrides reference existing skills
      for (const skillName of Object.keys(variant.skill_overrides)) {
        if (!variant.additional_skills.includes(skillName)) {
          // Check if it's in base role
          const baseRole = await this.loadRole(variant.base_role);
          if (!(skillName in baseRole.skills)) {
            errors.push(`Skill override references non-existent skill: ${skillName}`);
          }
        }
      }
    } catch (error) {
      errors.push(`Validation error: ${String(error)}`);
    }

    return errors;
  }
}

File: .claude/utils/variant_loader.py
"""
Variant Loader - Runtime skill combination mechanism for RAMS framework

This module provides functionality to load and instantiate role variants
by combining base roles with additional skills.
"""

import yaml
from pathlib import Path
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field


@dataclass
class SkillConfig:
    """Skill configuration"""
    name: str
    engine: Optional[str] = None
    tools: List[str] = field(default_factory=list)
    config: Dict[str, Any] = field(default_factory=dict)


@dataclass
class VariantConfig:
    """Variant configuration"""
    name: str
    base_role: str
    additional_skills: List[str] = field(default_factory=list)
    skill_overrides: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    skill_priority: Dict[str, int] = field(default_factory=dict)
    description: str = ""
    tags: List[str] = field(default_factory=list)


@dataclass
class RoleConfig:
    """Role configuration"""
    name: str
    skills: Dict[str, SkillConfig] = field(default_factory=dict)
    description: str = ""
    variants: List[str] = field(default_factory=list)


class VariantLoader:
    """Load and instantiate role variants"""
    
    def __init__(self, roles_dir: Path, skills_dir: Path):
        """
        Initialize variant loader
        
        Args:
            roles_dir: Path to roles directory
            skills_dir: Path to skills directory
        """
        self.roles_dir = Path(roles_dir)
        self.skills_dir = Path(skills_dir)
        self._role_cache: Dict[str, RoleConfig] = {}
        self._variant_cache: Dict[str, VariantConfig] = {}
    
    def load_role(self, role_name: str) -> RoleConfig:
        """
        Load base role configuration
        
        Args:
            role_name: Name of the role to load
            
        Returns:
            RoleConfig object
        """
        if role_name in self._role_cache:
            return self._role_cache[role_name]
        
        role_path = self.roles_dir / role_name / "role.md"
        if not role_path.exists():
            raise FileNotFoundError(f"Role not found: {role_name}")
        
        # Parse role.md to extract skills
        # This is a simplified implementation
        # In production, you'd parse the markdown properly
        role_config = RoleConfig(name=role_name)
        
        # Load variants if they exist
        variants_path = self.roles_dir / role_name / "variants.yaml"
        if variants_path.exists():
            variants = self._load_variants_yaml(variants_path)
            role_config.variants = [v.name for v in variants]
            for variant in variants:
                self._variant_cache[variant.name] = variant
        
        self._role_cache[role_name] = role_config
        return role_config
    
    def _load_variants_yaml(self, variants_path: Path) -> List[VariantConfig]:
        """Load variants from YAML file"""
        with open(variants_path, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)
        
        variants = []
        for variant_data in data.get('variants', []):
            variant = VariantConfig(
                name=variant_data['name'],
                base_role=variant_data['base_role'],
                additional_skills=variant_data.get('additional_skills', []),
                skill_overrides=variant_data.get('skill_overrides', {}),
                skill_priority=variant_data.get('skill_priority', {}),
                description=variant_data.get('description', ''),
                tags=variant_data.get('tags', [])
            )
            variants.append(variant)
        
        return variants
    
    def load_variant(self, variant_name: str) -> RoleConfig:
        """
        Load and instantiate a role variant
        
        Args:
            variant_name: Name of the variant to load
            
        Returns:
            RoleConfig object with merged skills
        """
        # Find variant in cache
        variant = self._variant_cache.get(variant_name)
        if not variant:
            # Try to find it by scanning all roles
            variant = self._find_variant(variant_name)
            if not variant:
                raise ValueError(f"Variant not found: {variant_name}")
        
        # Load base role
        base_role = self.load_role(variant.base_role)
        
        # Merge skills
        merged_skills = self._merge_skills(
            base_role.skills,
            variant.additional_skills
        )
        
        # Apply skill overrides
        for skill_name, override_config in variant.skill_overrides.items():
            if skill_name in merged_skills:
                skill = merged_skills[skill_name]
                for key, value in override_config.items():
                    setattr(skill, key, value)
        
        # Resolve conflicts based on priority
        resolved_skills = self._resolve_conflicts(
            merged_skills,
            variant.skill_priority
        )
        
        # Create new role config
        variant_role = RoleConfig(
            name=variant.name,
            skills=resolved_skills,
            description=variant.description,
            variants=[]
        )
        
        return variant_role
    
    def _find_variant(self, variant_name: str) -> Optional[VariantConfig]:
        """Find variant by scanning all roles"""
        for role_dir in self.roles_dir.iterdir():
            if role_dir.is_dir():
                variants_path = role_dir / "variants.yaml"
                if variants_path.exists():
                    variants = self._load_variants_yaml(variants_path)
                    for variant in variants:
                        if variant.name == variant_name:
                            self._variant_cache[variant.name] = variant
                            return variant
        return None
    
    def _merge_skills(
        self,
        base_skills: Dict[str, SkillConfig],
        additional_skill_names: List[str]
    ) -> Dict[str, SkillConfig]:
        """Merge base skills with additional skills"""
        merged = dict(base_skills)
        
        for skill_name in additional_skill_names:
            # Load skill configuration
            skill_path = self.skills_dir / skill_name / "SKILL.md"
            if skill_path.exists():
                # Create skill config (simplified)
                skill_config = SkillConfig(name=skill_name)
                merged[skill_name] = skill_config
            else:
                print(f"Warning: Skill not found: {skill_name}")
        
        return merged
    
    def _resolve_conflicts(
        self,
        skills: Dict[str, SkillConfig],
        priorities: Dict[str, int]
    ) -> Dict[str, SkillConfig]:
        """
        Resolve skill conflicts based on priority
        
        Higher priority values take precedence
        """
        # If no priorities, return as-is
        if not priorities:
            return skills
        
        # Sort skills by priority
        sorted_skills = sorted(
            skills.items(),
            key=lambda x: priorities.get(x[0], 0),
            reverse=True
        )
        
        return dict(sorted_skills)
    
    def list_variants(self, role_name: Optional[str] = None) -> List[str]:
        """
        List all available variants
        
        Args:
            role_name: If specified, only list variants for this role
            
        Returns:
            List of variant names
        """
        if role_name:
            role = self.load_role(role_name)
            return role.variants
        
        # List all variants
        all_variants = []
        for role_dir in self.roles_dir.iterdir():
            if role_dir.is_dir():
                variants_path = role_dir / "variants.yaml"
                if variants_path.exists():
                    variants = self._load_variants_yaml(variants_path)
                    all_variants.extend([v.name for v in variants])
        
        return all_variants
    
    def validate_variant(self, variant_name: str) -> List[str]:
        """
        Validate a variant configuration
        
        Args:
            variant_name: Name of the variant to validate
            
        Returns:
            List of validation errors (empty if valid)
        """
        errors = []
        
        try:
            variant = self._variant_cache.get(variant_name)
            if not variant:
                variant = self._find_variant(variant_name)
                if not variant:
                    errors.append(f"Variant not found: {variant_name}")
                    return errors
            
            # Validate base role exists
            base_role_path = self.roles_dir / variant.base_role / "role.md"
            if not base_role_path.exists():
                errors.append(f"Base role not found: {variant.base_role}")
            
            # Validate additional skills exist
            for skill_name in variant.additional_skills:
                skill_path = self.skills_dir / skill_name / "SKILL.md"
                if not skill_path.exists():
                    errors.append(f"Skill not found: {skill_name}")
            
            # Validate skill overrides reference existing skills
            for skill_name in variant.skill_overrides.keys():
                if skill_name not in variant.additional_skills:
                    # Check if it's in base role
                    base_role = self.load_role(variant.base_role)
                    if skill_name not in base_role.skills:
                        errors.append(f"Skill override references non-existent skill: {skill_name}")
            
        except Exception as e:
            errors.append(f"Validation error: {str(e)}")
        
        return errors


# Example usage
if __name__ == "__main__":
    # Initialize loader
    loader = VariantLoader(
        roles_dir=Path(".claude/roles/game-development"),
        skills_dir=Path(".claude/skills/game-development-role/skills")
    )
    
    # List all variants
    print("Available variants:")
    for variant in loader.list_variants():
        print(f"  - {variant}")
    
    # Load a specific variant
    try:
        variant_role = loader.load_variant("unity-gameplay-programmer")
        print(f"\nLoaded variant: {variant_role.name}")
        print(f"Description: {variant_role.description}")
        print(f"Skills: {list(variant_role.skills.keys())}")
    except Exception as e:
        print(f"Error loading variant: {e}")
    
    # Validate a variant
    errors = loader.validate_variant("unity-gameplay-programmer")
    if errors:
        print(f"\nValidation errors:")
        for error in errors:
            print(f"  - {error}")
    else:
        print("\nVariant is valid")
